api/spanner/push_run.go

```diff
@@ -73,38 +278,153 @@ func pushRun(ctx context.Context, api API, id int64) {
+		logger.Warningf("Spanner contains %d more rows than WPT report", -n)
 		return
 	}
+	if n == 0 {
+		logger.Infof("Run already uploaded to spanner")
+		return
+	}
+
+	log.Infof("Preparing data for run %d", run.ID)
+
+	ss := NewStructs()
+
+	r, err := NewRun(run)
+	if err != nil {
+		log.Errorf("Spanner push failed to constructing run: %v", err)
+		return
+	}
+	ss.AddRun(r)
+
+	for _, result := range report.Results {
+		if len(result.Subtests) == 0 {
+			t := NewTest(result.Test, nil)
+			ss.AddTest(t)
+			res := NewResult(result.Status, nil)
+			ss.AddResult(res)
+			ss.AddRunResult(r, res)
+			ss.AddRunResultTest(r, res, t, result.Message)
+		} else {
+			for _, s := range result.Subtests {
+				t := NewTest(result.Test, &s.Name)
+				ss.AddTest(t)
+				res := NewResult(s.Status, nil)
+				ss.AddResult(res)
+				ss.AddRunResult(r, res)
+				ss.AddRunResultTest(r, res, t, s.Message)
+			}
+		}
+	}
+
+	log.Infof("Generating row-based mutations for run %d", run.ID)
+	r1s, r2s, r3s, err := ss.ToMutations()
+	if err != nil {
+		log.Errorf("Spanner push failed to generating mutations: %v", err)
+		return
+	}
+	numRows := len(r2s) + len(r2s) + len(r3s)
+	log.Infof("Generated %d rows for run %d", numRows, run.ID)
+
+	log.Infof("Writing batches for %d-row run %d", numRows, run.ID)
+
+	s := semaphore.NewWeighted(numConcurrentBatches)
+	writeBatch := func(batchSync *semaphore.Weighted, rowGroupSync *sync.WaitGroup, rows []*spanner.Mutation, m, n int) {
+		defer rowGroupSync.Done()
+		defer batchSync.Release(1)
+		batch := rows[m:n]
 
-	shared.GetLogger(ctx).Infof("NOT IMPLEMENTED: Spanner push run would now push run if number of missing rows=%d > 0", n)
+		err := retry.Do(func() error {
+			log.Infof("Writing batch for %d-row run %d: [%d,%d)", len(rows), run.ID, m, n)
+
+			newCtx, cancel := context.WithTimeout(ctx, time.Second*60)
+			defer cancel()
+
+			_, err := sClient.Apply(newCtx, batch)
+			if err != nil {
+				log.Errorf("Error writing batch for %d-row run %d: %v", len(rows), run.ID, err)
+				return err
+			}
+
+			log.Infof("Wrote batch for %d-row run %d: [%d,%d)", len(rows), run.ID, m, n)
+			return nil
+		}, retry.Attempts(5), retry.OnRetry(func(n uint, err error) {
+			log.Warningf("Retrying failed batch batch for %d-row run %d: [%d,%d): %v", len(rows), run.ID, m, n, err)
+		}))
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	writeRows := func(rows []*spanner.Mutation) *sync.WaitGroup {
+		var wg sync.WaitGroup
+		var end int
+		for end = batchSize; end <= len(rows); end += batchSize {
+			wg.Add(1)
+			s.Acquire(ctx, 1)
+			go writeBatch(s, &wg, rows[0:], end-batchSize, end)
+		}
+		if end != len(rows) {
+			wg.Add(1)
+			s.Acquire(ctx, 1)
+			log.Infof("Writing small batch for %d-row run %d: [%d,%d)", len(rows), run.ID, end-batchSize, len(rows))
+			go writeBatch(s, &wg, rows[0:], end-batchSize, len(rows))
+			log.Infof("Wrote small batch for %d-row run %d: [%d,%d)", len(rows), run.ID, end-batchSize, len(rows))
+		}
+		return &wg
+	}
+
```

Comment:
Shall we use transactions here? Or is it too much data?

Reply:
IIRC this is way too much data for a transaction.